semeio/jobs/auto_correlate_observations/config.py

```diff
@@ -0,0 +1,398 @@
+        return lower_valid and upper_valid
+
+    return validator
+
+
+def _one_of(*valid_values):
+    msg = "Value must be one of: ({})".format(", ".join(map(str, valid_values)))
+
+    @cs.validator_msg(msg)
+    def validator(value):
+        return value in valid_values
+
+    return validator
+
+
+@cs.validator_msg("t must be an integer if a max cluster criteria is choosen")
+def _t_is_int_if_maxclust(spearman_config):
+    if spearman_config[_CRITERION] in (_MAXCLUST, _MAXCLUST_MONOCRIT):
+        return isinstance(spearman_config[_SPEARMAN_THRESHOLD], int)
+    return True
+
+
+@cs.transformation_msg(
+    (
+        "Will be defaulted to 5 for all max cluster criterias and "
+        "1.15 for all other criterias"
+    )
+)
+def _inject_default_t(config):
+    config = copy.deepcopy(config)
+    criterion = config.get(_CRITERION)
+    if _SPEARMAN_THRESHOLD in config:
+        pass
+    elif criterion in (_MAXCLUST, _MAXCLUST_MONOCRIT):
+        config[_SPEARMAN_THRESHOLD] = 5
+    elif criterion in (_INCONSISTENT, _DISTANCE, _MONOCRIT):
+        config[_SPEARMAN_THRESHOLD] = 1.15
+    else:
+        raise ValueError("Unknown criterion: {}".format(criterion))
+    return config
+
+
+_SPEARMAN_CORRELATION_SCHEMA = {
+    cs.MetaKeys.Type: cs.types.NamedDict,
+    cs.MetaKeys.Description: (
+        "The {sc} clustering is supporting multiple parameters."
+    ).format(sc=_SPEARMAN_CORRELATION),
+    cs.MetaKeys.Content: {
+        _FCLUSTER: {
+            cs.MetaKeys.Type: cs.types.NamedDict,
+            cs.MetaKeys.Description: (
+                "The {fcluster} implementation is backed by scipy and for a "
+                "more detailed description we refer the reader to the "
+                "documentation of scipy.cluster.hierarchy.fcluster` "
+                "(https://docs.scipy.org/doc/scipy/reference/generated"
+                "/scipy.cluster.hierarchy.fcluster.html)."
+            ).format(fcluster=_FCLUSTER),
+            cs.MetaKeys.ElementValidators: (_t_is_int_if_maxclust,),
+            cs.MetaKeys.Transformation: _inject_default_t,
+            cs.MetaKeys.Content: {
+                _SPEARMAN_THRESHOLD: {
+                    cs.MetaKeys.Type: cs.types.Number,
+                    cs.MetaKeys.Description: (
+                        "Scalar threshold for the clustering. When a "
+                        "'{maxclust}' {criterion} is used, the scalar gives "
+                        "the maximum number of clusters to be formed."
+                    ).format(maxclust=_MAXCLUST, criterion=_CRITERION),
+                    cs.MetaKeys.ElementValidators: (
+                        _bounds_validator(lower=0, lower_inclusive=False),
+                    ),
+                },
+                _CRITERION: {
+                    cs.MetaKeys.Type: cs.types.String,
+                    cs.MetaKeys.Description: (
+                        "The criterion to use in forming flat clusters. "
+                        "Defaults to {}."
```

Comment:
Name or not named parameters for formatting?

Reply:
I don't have any strong preferences. Do you?